syntax = "proto3";

package out;

option go_package = "../protos";

service VMManagementService {
  rpc StartVM (VMRequest) returns (VMResponse) {}
  rpc StopVM (VMRequest) returns (VMResponse) {}
  rpc DestroyVM (VMRequest) returns (VMResponse) {}
  rpc DestroyAllVMs (DestroyAllVMsRequest) returns (VMResponse) {}
  rpc ListAllVMs (ListAllVMsRequest) returns (ListAllVMsResponse) {}
  rpc ListVM (ListVMRequest) returns (ListVMResponse) {}
}

message VMRequest {
  string vm_name = 1;
  string entry_point = 2;
  // Next Id: 3.
}

message VMResponse {}

message DestroyAllVMsRequest {}

message ListAllVMsRequest {}

message ListVMRequest {
  string vm_name = 1;
  // Next ID: 2.
}

message ListVMResponse {
  VMInfo vm_info = 1;
  // Next ID: 2.
}

enum VMStatus {
  VM_STATUS_UNSPECIFIED = 0;
  VM_STATUS_RUNNING = 1;
  VM_STATUS_STOPPED = 2;
}

message VMInfo {
  string vm_name = 1;
  string ip = 2;
  string tap_device_name = 3;
  VMStatus status = 4;
  // Next Id: 5.
}

message ListAllVMsResponse {
  repeated VMInfo vms = 1;
  // Next Id: 2.
}

